### DISCUSSION TITLE
Old Examples On AI Deck Not Being Overwritten

### DISCUSSION THREAD

**Discussion by LucasTheIntern** [bitcraze/discussions#1512] [Q&A]
Hey team, 

I'm having an issue very similar to (#542), (#637), and ([issue #126](https://github.com/bitcraze/aideck-gap8-examples/issues/126)) where I have flashed a customized example script to the GAP8 on the AI deck, and now cannot seem to overwrite it. 

I don't _think_ I made the mistake of sending cpx packets with no delay, but it is possible. The last version of the example that I flashed does include a delay, but it might not have always been there since I did make some changes. It took me waaaay too long to notice that newer versions of my example weren't being flashed properly because I was more focused on the Crazyflie's STM32. I only noticed when I tried to modify the behavior of the heartbeat task (adopted from the wifi streamer example), and inspection of the console output confirmed that there were a lot of debugging print statements being executed that are no longer present in the code. 
____________________________________________________________________________________________________________________________________________________

My first thought was to try reflashing the default firmware from the CF Client connect>bootloader GUI. It seems to behave normally after that, but restarting the crazyflie causes it to resume running the old code. 
____________________________________________________________________________________________________________________________________________________

I then tried flashing the bootloader over JTAG (later found #542 which seems to suggest the same thing as the answer). Again, it flashes successfully, and the old heartbeat task ceases. But on restarting, it resumes the old code again. Just in case I'm messing something up or missing something, here are the commands I'm using to do that:
```
//IN TERMINAL:

git clone https://github.com/bitcraze/aideck-gap8-bootloader.git

//(This will cause an error if you have already run it on this VM, but everything is fine ;)

cd aideck-gap8-bootloader

docker run --rm -it -v $PWD:/module/ --device /dev/ttyUSB0 --privileged -P bitcraze/aideck /bin/bash -c 'export GAPY_OPENOCD_CABLE=interface/jlink.cfg; source /gap_sdk/configs/ai_deck.sh; cd /module/;  make all image flash'
```
____________________________________________________________________________________________________________________________________________________

I've also tried flashing the newest version of my code Over-The-Air. That results in the LED it should be using for the heartbeat task (LED 2) to get stuck "ON", which is not at all expected behavior, and I don't see of my new print statements coming from the GAP8 in the console, though I do see others from the ESP32. 
Over-The-Air flashing commands:
```
cd aideck-gap8-examples

docker run --rm -v ${PWD}:/module --privileged bitcraze/aideck tools/build/make-example examples/MyExamples/AIIAF clean all

cfloader flash examples/MyExamples/AIIAF/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/target.board.devices.flash.img deck-bcAI:gap8-fw -w radio://0/80/2M/E7E7E7E7E7
```

I have ensured, of course, that no other decks are attached during these attempts. 
____________________________________________________________________________________________________________________________________________________

Finally, I have tried flashing my example over JTAG using the following commands based on ([flashing documentation](https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/infrastructure/flashing/)) and the solution by hmllr and bhabas in (#637):

```
cd aideck-gap8-examples

docker run --rm -it -v $PWD:/module/ --device /dev/ttyUSB0 --privileged -P bitcraze/aideck tools/build/make-example examples/MyExamples/AIIAF flash
```
However, this results in "ERROR 1," which claims: 
```
Error: no device found
Error: unable to open ftdi device with vid 15ba, pid 002a, description 'Olimex OpenOCD JTAG ARM-USB-TINY-H', serial '*' at bus location '*'
```

This reminds me of (#852), which was related to flashing the bootloader over JTAG. I did have that problem before, and you'll see I have made that correction in the process above for flashing the bootloader. I'm not sure if there is a way to specify the .cfg file needed in this command. But, curiously enough, I can run the following commands with no trouble at all:
```
cd aideck-gap8-examples

docker run --rm -it -v $PWD:/module/ --device /dev/ttyUSB0 --privileged -P bitcraze/aideck tools/build/make-example examples/MyExamples/AIIAF clean all
```
The only difference is that this ends with "clean all" instead of "flash." The terminal output looks like it is flashing:
```
gapy --target=ai_deck --platform=board --work-dir=/module/examples/MyExamples/AIIAF/BUILD/GAP8_V2/GCC_RISCV_FREERTOS    run --image --binary=/module/examples/MyExamples/AIIAF/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/AIIAF 
gapy --target=ai_deck --platform=board --work-dir=/module/examples/MyExamples/AIIAF/BUILD/GAP8_V2/GCC_RISCV_FREERTOS    run --flash --binary=/module/examples/MyExamples/AIIAF/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/AIIAF 
gapy --target=ai_deck  gen_flash_image --boot-loader /module/examples/MyExamples/AIIAF/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/AIIAF  -o /module/examples/MyExamples/AIIAF/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/flash.img 
```
But doing so seems to have no effect whatsoever on the AI deck. No change in LED 2 behavior, and no change in console output. 
____________________________________________________________________________________________________________________________________________________

Needless to say, I'm stumped at this point. My best guess is that I'm missing something from the answers on (#542) and (#637), but I don't know what that might be. I would greatly prefer to flash my examples over JTAG if possible, as it seems to be orders of magnitude faster than OTA... at least on my machine. 

Any and all insight is much appreciated!!!

**Comment by gemenerik** [comment]
OTA flashing relies on CPX working, not sure if it should be possible to overload it with your app

LED stuck might be related to GAP8 initialization problems. Seeing as you are actively developing on the deck, I recommend implementing the workaround with a resistor https://github.com/bitcraze/aideck-gap8-examples/issues/105

As you can read in your error, the docker tries to flash the GAP8 with the `Olimex OpenOCD JTAG ARM-USB-TINY-H`, while in your first code block you are adding `export GAPY_OPENOCD_CABLE=interface/jlink.cfg` as an argument, so I assume you are trying to use the jlink instead. Perhaps the easiest is to modify your aideck-with-autotiler docker image's `tools/make-example` file. By default it has `export GAPY_OPENOCD_CABLE=interface/ftdi/olimex-arm-usb-tiny-h.cfg`, change this to the jlink.

Let me know if this helps.
